### ISSUE TITLE
reusing map values as map keys

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bcc#347]
Low priority, but I've found that map values can't be reused as keys. Eg:

``` C
struct info_t {
    u32 pid;
    char name[TASK_COMM_LEN];
};
BPF_HASH(infobyreq, struct request *, struct info_t);
BPF_HASH(counts, struct info_t, u64);
[...]
infop = infobyreq.lookup(&req);
valp = counts.lookup_or_init(infop, &zero);
```

Gives:

``` Shell
from 17 to 43: R0=map_value_or_null(ks=8,vs=20) R6=map_value(ks=8,vs=20) R10=fp
43: (18) r1 = 0x2f70f300
45: (bf) r2 = r6
46: (85) call 1
R2 type=map_value expected=fp
```

I can __builtin_memcpy() as a workaround to a new key.

**Comment by pchaigno** [comment]
@4ast Would allowing map helpers access to map element values in the verifier (in the vein of https://patchwork.ozlabs.org/patch/712889/) be a sensible fix for this? If so, I have a patch ready, which I can submit to the iovisor mailing list.

**Comment by 4ast** [comment]
there was quite a bit work on the verifier side recently and helpers should be able to access map values already. What's missing? and yes, the patch would be great :)

**Comment by pchaigno** [comment]
In the 4.13-rc7 version, other helpers appear to be able to access map values, but not map helpers. I'll try to send the patch today. Best to discuss it over code.

**Comment by pchaigno** [comment]
I just posted the patch on the iovisor mailing list.

**Comment by matthew-olson-intel** [comment]
@pchaigno Since that patch seems to be merged, do we need to re-test and confirm that this issue is resolved?

**Comment by pchaigno** [comment]
I believe this can be closed. My patch was released in Linux 4.18.

**Comment by brendangregg** [comment]
Thanks @pchaigno and @matthew-olson-intel .
